src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/ProvenanceHeaderConditionalCreateHandler.cs

```diff
@@ -0,0 +1,76 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Serialization;
+using MediatR;
+using Microsoft.Health.Fhir.Api.Features.Exceptions;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Messages.Create;
+using Microsoft.Health.Fhir.Core.Messages.ProvenanceHeader;
+using Microsoft.Health.Fhir.Core.Messages.Upsert;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources
+{
+    /// <summary>
+    /// Handler for conditional create request with "X-Provenance header".
+    /// Conditionally creates resource  and Provenance resource provided in "X-Provenance" header with it's target as resource.
+    /// </summary>
+    public sealed class ProvenanceHeaderConditionalCreateHandler : IRequestHandler<ProvenanceHeaderConditionalCreateRequest, UpsertResourceResponse>
+    {
+        private readonly FhirJsonParser _fhirJsonParser;
+        private readonly IMediator _mediator;
+
+        public ProvenanceHeaderConditionalCreateHandler(FhirJsonParser fhirJsonParser, IMediator mediator)
+        {
+            EnsureArg.IsNotNull(fhirJsonParser, nameof(fhirJsonParser));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+
+            _fhirJsonParser = fhirJsonParser;
+            _mediator = mediator;
+        }
+
+        public async Task<UpsertResourceResponse> Handle(ProvenanceHeaderConditionalCreateRequest request, CancellationToken cancellationToken)
```

Comment:
I would have probably explored a completely different approach :P, something along the lines of:

- Create a new PipelineBehavior that intercepts Create/Upsert requests
- Inject IMediator, IRequestContext 
- Check IRequestContext for ProvenanceHeader - then this similar logic

Reply:
Yeah, I'm quite unhappy with 4 different handles which all do the same thing.
My thoughts was to unite all 4 original requests classes into one with parameter which specifies behavior, and have one handler to rule them all and one behavior pipeline.

IRequestContext should do the trick, but it's somehow not pure enough for my taste (coming from person who copypaste 4 classes) it's some side thing related to asp.net pipeline rather than pure classes which we send to mediator.
I'll give it a try and update PR.